lib/persisted-model.js

```diff
@@ -651,6 +673,22 @@ module.exports = function(registry) {
 
     setRemoting(PersistedModel, 'replaceOrCreate', replaceOrCreateOptions);
 
+    setRemoting(PersistedModel, 'upsertWithWhere', {
+      aliases: ['patchOrCreateWithWhere'],
+      description: 'Update an existing model instance or ' +
+      'insert a new one into the data source based on the ' +
+      'where criteria.',
+      accessType: 'WRITE',
+      accepts: [
+        { arg: 'where', type: 'object', http: { source: 'query' },
+          description: 'Criteria to match model instances' },
+        { arg: 'data', type: 'object', http: { source: 'body' },
+          description: 'An object of model property name/value pairs' },
+      ],
+      returns: { arg: 'data', type: typeName, root: true },
+      http: { verb: 'put', path: '/upsertWithWhere' },
```

Comment:
@mountain1234585 Lets got for `POST /upsertWithWhere` and please make sure you add enough test coverage; please look into [PR#2316](https://github.com/strongloop/loopback/pull/2316) which gives you an idea what test coverage should be added in this patch.